### ISSUE TITLE
Configuring JMX connection

### ISSUE THREAD

**Issue by dmuelle** [OpenLiberty/docs#1695]
Liberty supports two JMX connectors: local connector and REST connector. Each connector is enabled through a different Liberty feature.

SME: Don Bourne

Combine the following config tasks into a single task topic

-https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_admin_localconnector.html   
- https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_admin_restconnector.html

incorporate info from the container topic as needed

https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_admin_jmx.html

provide a link to https://draft-openlibertyio.mybluemix.net/docs/ref/general/#jmx-metrics.html for more info on JMX, when to use & why

**Comment by ManasiGandhi** [comment]
Link to the draft https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/configuring-jmx-connection.html

**Comment by donbourne** [comment]
@navidsh , could you please review this doc?  I see you guys own JMX in Liberty, so in better position to be able to comment.

**Comment by navidsh** [comment]
Hi @ManasiGandhi,

Looks good overall. Just a few comments:

- [ ] It would be good to include info in the "Restriction" box from https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_admin_jmx.html. We get customers having the same issue mentioned there.

- [ ] Please also include the following node: "An application deployed on Liberty has unrestricted access to its MBeanServer directory."

- [ ] "Enable the `transportSecurity-1.0` Liberty feature": There are some details mentioned in "Configure SSL certificates in the server.xml file." in [Configure SSL certificates](https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_admin_restconnector.html) needed for setting up SSL on the server. Is there an equivalent page on the Open Liberty side you could have a reference to in the doc?

- [ ] "Configure a user or group to the administrator...": Similarly, is there a page equivalent to [Map to the administrator role for Liberty](https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_sec_mapadmin.html?view=kc) you could refer to from the doc?

- [ ] "...collective routing, and collective deployment." Please remove mentions of collective as collective features are not available in Open Liberty.

- [ ] "Access the REST connector" some details from KC are missing here. My question is if this whole doc on Open Liberty website is supposed to be a copy of the content from KC or just a summary of KC articles?

**Comment by ManasiGandhi** [comment]
@navidsh I worked on your review https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/configuring-jmx-connection.html
- [x] It would be good to include info in the "Restriction" box from https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_admin_jmx.html. We get customers having the same issue mentioned there.

- [x] Please also include the following node: "An application deployed on Liberty has unrestricted access to its MBeanServer directory."

- [x] "Enable the `transportSecurity-1.0` Liberty feature": There are some details mentioned in "Configure SSL certificates in the server.xml file." in [Configure SSL certificates](https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_admin_restconnector.html) needed for setting up SSL on the server. Is there an equivalent page on the Open Liberty side you could have a reference to in the doc?

- [x] "Configure a user or group to the administrator...": Similarly, is there a page equivalent to [Map to the administrator role for Liberty](https://www.ibm.com/support/knowledgecenter/SSAW57_liberty/com.ibm.websphere.wlp.nd.multiplatform.doc/ae/twlp_sec_mapadmin.html?view=kc) you could refer to from the doc?

- [x] "...collective routing, and collective deployment." Please remove mentions of collective as collective features are not available in Open Liberty.

- [x] "Access the REST connector" some details from KC are missing here. My question is if this whole doc on Open Liberty website is supposed to be a copy of the content from KC or just a summary of KC articles?

**Comment by navidsh** [comment]
- [x] Formatting issues across [the rendered page](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.10/configuring-jmx-connection.html) e.g.
![image](https://user-images.githubusercontent.com/2364853/94273213-1846aa80-ff12-11ea-9668-0336a51717c9.png)
- [x] I don't think users can enable features as follows `<feature>restConnector-2.0/>`. It's usually in this way:
```xml
<feature>restConnector-2.0</feature>
```
- [x] Having both of these paragraphs seem redundant: 
  > You can configure and access the REST connector on Open Liberty with the following steps:
  > ...
  > You can configure and access the REST connector on Open Liberty with the following:

- [x] Transport security section: the link should point to a doc about transport security. Currently it's pointing to Application Security
- [x] I'd suggest to keep formatting consistent when referring to directories. e.g. "...which is in `wlp/clients/restConnector.jar `file and..." vs. "...the clients/restConnector.jar file."
- [x] " Use -J flags" -> " Use `-J` flags"?

**Comment by ManasiGandhi** [comment]
@navidsh Here's the new link to the draft with the changes https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/configuring-jmx-connection.html.

**Comment by navidsh** [comment]
- "Access the REST connector directly by using an HTTPS call. You can see the full list of REST API endpoints, along with information on how to use the APIs, by enabling the API Discovery feature on an Open Liberty server."

This needs to be reworded since the API Discovery feature isn't available in Open Liberty. (I missed it in the last review).:

Suggested change:
> Besides using a JMX client application or JConsole tool, you can make direct HTTPS calls to the server.

The rest of the changes look good.

**Comment by ManasiGandhi** [comment]
@navidsh Updated link with the latest changes https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/configuring-jmx-connection.html

**Comment by navidsh** [comment]
@ManasiGandhi The latest looks good to me. Thank you.

**Comment by ManasiGandhi** [comment]
Updated draft link https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/configuring-jmx-connection.html

**Comment by lauracowen** [comment]
Thanks. My following comments relate mainly to the overall structure and organisation of the info. Can you make sure, when you've addressed them, that the whole topic makes logical sense as a whole and is clear up-front why the reader should continue reading past the first sentence - what value it provides them, etc.

- Intro
  - Don't think you need to give the feature names in the opening paragraph - the "how" comes later. Focus on the what and why at this stage.
  - Can you add something to say what a connector (in this context) is and why someone might want to use it (basically why this information and the connectors might be relevant to them so that they should carry on reading)? Later in the topic it says things like "access through the local connector" but it's not clear what you might be accessing or why. In fact, if you take a look at the generated feature doc topics for each of these connectors, they do explain what the connectors are, so take a look.
  - It's also not clear why you would use one rather than the other. While the name makes it clear that the local connector is used for a local JMX connection, it's not clear why you have to switch to the rest connector for the secure JMX connection. The intuitive alternative to "local connection" would be "remote connection", not "secure connection" - should the second section be "Configuring a secure remote JMX connection to Open Liberty" or something like that (check with SME)?
- Topic structure - can you apply the [task topic guidance](https://github.com/OpenLiberty/docs/wiki/How-to-write-task-topics) that @dmuelle has put together; eg don't us the "About this task" or "Procedure" headings that are in the KC/DITA topics. But maybe chat with David if you need more info about how to apply it here.
- The headings need either to be plural (eg "Configuring JMX connections") or have a definitive article inserting: "Configuring a JMX connection", "Configuring a local...", "Configuring a secure..." Probably plural, at least for the main title. Maybe see what Charlotte/David have done with other similar articles.
- "local connection" section.
  - If you remove the "About this task" heading, I don't think you necessarily need the "You can access" sentence at the start, unless you can make it clearer what you might be accessing using the local connector.
  - In fact, if you remove the whole of the "About this task" section, instead have a paragraph in the intro to the topic (after your first paragraph) that explains what the local connector is and why you'd use it, and then what the REST connector is and why you'd use it - talk to SME but start by drafting something based on the introductory paragraphs in the generated feature doc topics for these connectors. That bit about SDK is explained better in the generated topic I think.
  - There's a dodgy backtick "`" in Step 1.
  - Don't include steps on how to enable features in topics - that is already available in the generated feature doc ([local connector](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.12/reference/feature/localConnector-1.0.html) and [rest connector](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.12/reference/feature/restConnector-2.0.html)). Instead, just say to enable it and link to the relevant feature doc topic (check/ask what Charlotte has done in terms of wording in other topics).
  - Karen will pick up on "by the following:" - can you reword?
  - If JConsole and JMX client are alternatives, they shouldn't be consecutive numbered steps. I would suggest maybe not using numbered steps in this section at all and just having two separate paragraphs or even sub-sections. You could have a short paragraph/sub-section on JConsole and how you can copy the contents of the ...directory into the Remote Process field, then click Connect. And then have a short paragraph/sub-section on JMX client and how you can copy the contents of the ... directory into the constructor etc. I don't think you need to instruct people on how copy and paste works with clipboards - I think we can assume they understand that.
  - Are you accessing the local connector or are you accessing something by using the local connector? Can you check?

- "secure JMX connection" section
  - As above re the "About this task" section.
  - Enable the features - point them to the relevant generated feature doc topics, don't duplicate that info here. Similarly, the TLS generated doc contains an example of how to specify a keystore so point them to that instead of duplicating it here (there's no point when it's that simple a config).
  - Step 3 - good to point them to the Application Security topic but can you point them to the specific section that contains the role configuration examples?
  - Step 4 reads like introductory info about the rest connector and should probably be moved up to the new introduction to the topic itself? I'm not sure it goes here.
  - Similarly the rest of the topic seems more like introductory info than info that should be at the end (also has some dodgy capitalisation in the first sentence). But make sure that the whole topic makes logical sense (eg some of the last paragraph maybe shouldn't go at the top?).
- Then at the end of each of the two sections, I think there needs to be a summing up section re-stating what the user has just done and what it means for them.

**Comment by ManasiGandhi** [comment]
@lauracowen I worked on your review for the JMX topic. Here is the updated draft link https://draft-openlibertyio.mybluemix.net/docs/20.0.0.12/configuring-jmx-connection.html



    Intro
        

- [x] Don't think you need to give the feature names in the opening paragraph - the "how" comes later. Focus on the what and why at this stage.

       

- [x]  Can you add something to say what a connector (in this context) is and why someone might want to use it (basically why this information and the connectors might be relevant to them so that they should carry on reading)? Later in the topic it says things like "access through the local connector" but it's not clear what you might be accessing or why. In fact, if you take a look at the generated feature doc topics for each of these connectors, they do explain what the connectors are, so take a look.

        

- [x] It's also not clear why you would use one rather than the other. While the name makes it clear that the local connector is used for a local JMX connection, it's not clear why you have to switch to the rest connector for the secure JMX connection. The intuitive alternative to "local connection" would be "remote connection", not "secure connection" - should the second section be "Configuring a secure remote JMX connection to Open Liberty" or something like that (check with SME)?

    

- [ x] Topic structure - can you apply the task topic guidance that @dmuelle has put together; eg don't us the "About this task" or "Procedure" headings that are in the KC/DITA topics. But maybe chat with David if you need more info about how to apply it here.

    The headings need either to be plural (eg "Configuring JMX connections") or have a definitive article inserting: 

- [x] "Configuring a JMX connection", "Configuring a local...", "Configuring a secure..." Probably plural, at least for the main title. Maybe see what Charlotte/David have done with other similar articles.

    "local connection" section.
        

- [x] If you remove the "About this task" heading, I don't think you necessarily need the "You can access" sentence at the start, unless you can make it clearer what you might be accessing using the local connector.

       

- [x]  In fact, if you remove the whole of the "About this task" section, instead have a paragraph in the intro to the topic (after your first paragraph) that explains what the local connector is and why you'd use it, and then what the REST connector is and why you'd use it - talk to SME but start by drafting something based on the introductory paragraphs in the generated feature doc topics for these connectors. That bit about SDK is explained better in the generated topic I think.

       

- [x]  There's a dodgy backtick "`" in Step 1.

        

- [x] Don't include steps on how to enable features in topics - that is already available in the generated feature doc (local connector and rest connector). Instead, just say to enable it and link to the relevant feature doc topic (check/ask what Charlotte has done in terms of wording in other topics).

        

- [x] Karen will pick up on "by the following:" - can you reword?

        

- [x] If JConsole and JMX client are alternatives, they shouldn't be consecutive numbered steps. I would suggest maybe not using numbered steps in this section at all and just having two separate paragraphs or even sub-sections. You could have a short paragraph/sub-section on JConsole and how you can copy the contents of the ...directory into the Remote Process field, then click Connect. And then have a short paragraph/sub-section on JMX client and how you can copy the contents of the ... directory into the constructor etc. I don't think you need to instruct people on how copy and paste works with clipboards - I think we can assume they understand that.

        

- [x] Are you accessing the local connector or are you accessing something by using the local connector? Can you check?

    "secure JMX connection" section
        

- [x] As above re the "About this task" section.

        

- [x] Enable the features - point them to the relevant generated feature doc topics, don't duplicate that info here. Similarly, the TLS generated doc contains an example of how to specify a keystore so point them to that instead of duplicating it here (there's no point when it's that simple a config).

        

- [x] Step 3 - good to point them to the Application Security topic but can you point them to the specific section that contains the role configuration examples?

        

- [x] Step 4 reads like introductory info about the rest connector and should probably be moved up to the new introduction to the topic itself? I'm not sure it goes here.

        

- [x] Similarly the rest of the topic seems more like introductory info than info that should be at the end (also has some dodgy capitalisation in the first sentence). But make sure that the whole topic makes logical sense (eg some of the last paragraph maybe shouldn't go at the top?).

   

- [x]  Then at the end of each of the two sections, I think there needs to be a summing up section re-stating what the user has just done and what it means for them.

**Comment by lauracowen** [comment]
Thanks, this is looking better at a high-level structure. I think the next step is to really clarify the point of the connectors and why you'd use them. That will help you to simplify the language used to explain them.

- I'm still not clear (as a reader) what the purpose of these connectors is, and why I might choose one or the other. The intro needs some concrete examples of why you might use JMX connectors at all. And then needs concrete examples of why you might choose the local vs rest connector. You'll need to ask the SME for this, such as what might you want to monitor an application for by using a connector? The explanations read like you don't really know and you're fluffing it with words that sound about right but aren't that meaningful. The explanations need to sound like you do know what you're talking about (even though you don't/shouldn't have deep technical knowledge that the SME has). Ask Fred to explain what connectors are for so that you understand the basic concepts of what is connected to what and why an Open Liberty application might want to connect to something else using one.

- Avoid repeating the same thing in different sentences without adding value; eg "Therefore, the localConnector feature can be used only for local connections." says much the same thing as the previous sentence, just more vaguely.

- "The restConnector feature provides a direct and reliable method to access the JVM MBeanServer, either locally or remotely." - but why? (also check the spelling of "MBeanServer")

- When you're talking generally about feature names, use the human-readable name, not the machine-readable name. So "With the Admin Local Connector feature...". Check with David/Charlotte for the guidance but I think it might be to include the machine-readable name in parentheses after it on first use. So "With the Admin Local Connector feature (localConnector)..." but check also when you should be using the macros to include links or not (I can't remember what was decided).

- The headings for the sections still need either an article ("a" or "the") or to be pluralised as they're not grammatically correct at the moment. Maybe check with Karen which it should be for headings within topics.

- Go through and proofread and check the grammar (or ask someone else to). There are a lot of typos or small grammatical errors throughout.

- I know I asked for a summary sentence but I don't think this adds much value: "Thus, you can access the local JMX connector on Open Liberty and manage resources." because it doesn't actually summarise anything - it just assumes the reader has understood the previous paragraphs then, in effect, says "so there you go (and if you didn't understand, tough)". Leave this for now until you have a clearer idea (and shown in the intro) of what the user achieves by using the connectors in their application.

Honestly, I don't know very much about using JMX connectors with Open Liberty either so I'm not expecting you to become an expert at all. However, I should be able to gain a basic understanding of the point of this task just by reading your topic, which I currently can't get. I think the challenge here is asking the right questions of the SME - chat with Charlotte or David about what questions you could ask if you need some inspiration.

**Comment by lauracowen** [comment]
Some other notes from Alasdair (commenting on the existing KC topics but relevant here too and catching things I didn't think of but we really should ensure are clear):

- Make clear to the reader why you would use JMX versus anything else (what are the alternatives?). Don't assume they know.
- Make sure the information is provided in the context of containerised apps (eg in Docker containers). That's something I don't know so you'll need the SME to help with checking through to make sure it makes sense in a containerised world. For example, what does "local" and "remote" mean in the context of containers (state it in those terms)? What does it mean to manage resources when they might be in different containers when deployed to production, and would you do that interactively (logging into the running Docker container) once deployed or is it more likely that you'd set things up in advance in the Docker image? Why would you interactively manage resources in a container when you can/should just create a new container with new settings and throw the old one away?

In addition to Fred, you might need to talk to Arthur de Magalhaes (or some other cloud/container SME) to understand if/how this would be used for real in a container-based cloud (such as IBM Cloud).

**Comment by lauracowen** [comment]
Okay, through another conversation with Alasdair...

- don't talk about local connector at all. It doesn't have a purpose in a cloud/container environment. Its purpose is to communicate between apps in the same JVM/machine (with a container, you wouldn't have more than one app in a container). Mostly local connector is available for historical reasons.
- remote connector - it's there for if you want to do remote admin things eg read performance data. But you can also read performance data using the MP Metrics feature. A lot of monitoring products use JMX to get stats (even though you can do that with MicroProfile as well). It's one of those situations where there are multiple ways to do the same thing. So, I think talk to your SME and work out a topic that just focuses on the remote connector and see if you can get some reasons you can mention for why a user would use JMX versus (say) MicroProfile Metrics to get the data need (and give examples of the types of data you can get from each). See if there are other useful uses of the remote connector that are relevant beyond reading performance data that are relevant to cloud/containers (eg you can start and stop apps with the connector but that's not really relevant because you would just start/stop the container really).

BTW, to be clear, the target user for this info is someone doing admin, not an app developer.

**Comment by ManasiGandhi** [comment]
I have been following up with the SME. He (Navid) is moving to a new organization and is in the process of knowledge transfer. Laura mentioned some new points in the issue and also Alasdair's suggestions. Navid said he will need to discuss those points with his team and needs some time to get back to me. So waiting on that.

**Comment by Charlotte-Holt** [comment]
@ManasiGandhi Maybe when you get back, we can look at this one together. As I understand it, I think we could talk about setting up a secure JMX connection (restConnector) in the context of gathering JMX metrics. I have a couple of topics that discuss JMX metrics, so let's circle back on this one together when we can.

**Comment by ManasiGandhi** [comment]
Updated link to the draft https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/configuring-jmx-connection.html

**Comment by ManasiGandhi** [comment]
@pgunapal Hi Prashanth. Here are some of my questions regarding the JMX connection topic (draft link- https://draft-openlibertyio.mybluemix.net/docs/21.0.0.2/configuring-jmx-connection.html)
- Why you would use JMX versus anything else ,what are the alternatives?
- How do JMX connectors work in the context of containerised apps, eg in Docker containers?
For example, what does "local" and "remote" mean in the context of containers?
- What does it mean to manage resources when they might be in different containers when deployed to production, and would you do that interactively by logging into the running Docker container once deployed or is it more likely that you'd set things up in advance in the Docker image?
- Why would you interactively manage resources in a container when you can/should just create a new container with new settings and throw the old one away?
- How this would be used for real in a container-based cloud such as IBM Cloud?

**Comment by pgunapal** [comment]
@leochr @idlewis will be the best suited to answer these questions, as they are owners for Containers and JMX Rest connectors.

**Comment by leochr** [comment]
@ManasiGandhi I am not sure what'll be the alternative. Manipulating the container after it's deployed goes against the container best practices. So yes, an application image should be built with the necessary artifacts.

**Comment by idlewis** [comment]
"Your application can remotely connect...."
This sentence is misleading I think. Normally, 'your application' would be running on the liberty server, and would be the thing that is being managed.
I might just say 'You can remotely connect...'

"An application deployed on Open Liberty has unrestricted access to its MBeanServer directory"
This sentence seems out of place.  It is unrelated to how you configure the connection.
Instead it is describing what you can do with JMX. I can't see how it relates to anything else
in the article. I might leave it out.

"HTTPS calls use the server-side of the connector with any programming language that can make HTTPS calls, 
such as C++, JavaScript, and curl, to use the REST APIs. A Java client uses the client-side of the connector, 
which is in the wlp/clients/restConnector.jar file and implements the javax.management.MBeanServerConnection"
These two sentences also seems out of place. They are not describing authentication/authorization of the connection which is what the rest of the article seems to focus on. If they are left in, maybe put them in the first paragraph, which seems to be a more general description of JMX.


"Configure a user or group to the administrator role"
This section is almost completely about security config rather than JMX specifically.
I don't know enough about security to review this, you should probably get someone else to have a look

**Comment by dmuelle** [comment]
## Peer review

I think the topic is looking good. However,  too much of it is focused on role configuration, which we mostly document elsewhere and the topic doesn't tell you how to access the rest connector. I recommend making the main section of the topic closer to the procedure laid out in the IBM docs topic [Configuring secure JMX connection to Liberty](https://www.ibm.com/docs/en/was-liberty/nd?topic=jmx-configuring-secure-connection-liberty). And the role config that is not already documented elsewhere should be moved to more general locations. It is relevant to JMX but not exclusively; users shouldn't have to look at this topic to find role config info for unrelated purposes (such as other, non-JMX related REST APIs)

### short description
suggest revising this to focus on user goals rather than tooling:

`You can monitor and manage registered resources in your Open Liberty server by using Jakarta Management Extensions (JMX) management beans (MBeans). You can access individual JMX MBeans by remotely connecting to the Open Liberty MBean server through a secure JMX connection to the Open Liberty REST connector.`


### Configuring a secure remote JMX connection to Open Liberty

I don't think we need to mention SSL at all. Though the generated documentation for rest connector only says that it enables the SSL feature, that feature is superseded by the TLS feature- the generated doc for that feature indicates that it is enabled by the rest-connector feature. We may need to open an issue against the metatype to get this updated but we dont want to point anyone to a superseded feature. Plus all the relevant config info is on the TLS feature page.


`You can get a secure JMX connection for your application remotely with the Admin REST Connector feature by using the TLS protocol. Although the SSL protocol is now deprecated, the term SSL is often used as a synonym for TLS. The REST connector feature includes the Secure Socket Layer feature.`

--->

`You can configure a secure JMX connection by enabling the Admin REST Connector feature, configuring Transport Layer Security (TLS), and adding at least one user to the administrator or reader management roles. By default, the Admin REST Connector feature enables the Transport Security feature, which supports TLS connections. Remote access with a JMX connection is protected by a single administrator or reader role through the HTTPS port that is defined by the default httpEndpoint element. `

`You can access an Open Liberty REST connector from a Java client or directly through an HTTPS call to the JMX endpoints of the administrative REST APIs. A Java client uses the client-side of the REST connector, which is in the `wlp/clients/restConnector.jar` file and implements the `javax.management.MBeanServerConnection` interface. HTTPS calls use the server-side of the REST connector. Any programming language that can make HTTPS calls can use the REST APIs.` 

After that, I would replace the remainder of the section with the four steps from the KC article, but adapt them for open liberty.  

To configure a secure JMX connection, complete the following steps:

1. Enable the [Admin REST connector feature] and  [Configure TLS certificates](https://www.openliberty.io/docs/21.0.0.10/reference/feature/transportSecurity-1.0.html) in your server.xml file. 
Ensure that the CN value of the certificate subjectDN entry is the host name of the machine where the server is running. If you plan to use the  jConsole tool that the trust store contains the certificate of the server in the jConsole connection.

2. (https://www.openliberty.io/docs/21.0.0.10/reference/feature/appSecurity-3.0.html#_configure_rest_api_management_roles)[Configure at least one user or group to the administrator or reader role] in your server.xml file. For testing and development purposes, you can quickly configure a single user in the administrator role by [configuring QuickStart security](https://www.openliberty.io/docs/21.0.0.10/reference/feature/appSecurity-3.0.html#_configure_a_basic_user_registry_with_quickstart_security).

3. Access the REST connector by using a Java client application, through [the jConsole tool](https://docs.oracle.com/en/java/javase/17/management/using-jconsole.html#GUID-77416B38-7F15-4E35-B3D1-34BFD88350B5) that is provided by the Java SDK, or by making HTTPS calls directly to the server.

See Also
- [JMX metrics](https://www.openliberty.io/docs/21.0.0.10/introduction-monitoring-metrics.html#_jmx_metrics)

----
### Configure a user or group to the administrator role

I dont think we need any of the role mapping stuff on this page. The quick start and basic mapping is already on the appsecurity page. The LDAP example should be on the ldapregtistry page. The access ID stuff probably belongs with the other role mapping info on the appSecurity page, but it needs more context.

**Comment by ManasiGandhi** [comment]
@dmuelle I worked on your review. Here is a link to the draft https://draft-openlibertyio.mybluemix.net/docs/21.0.0.11/configuring-jmx-connection.html.

**Comment by dmuelle** [comment]
Looks good, a few minor things
- [x] provide a link for the Admin REST Connector feature and the Transport Security feature on first mentions
- [x] link to httpEndpoint config doc on first mention, and remove monospace from linked text
- [x] in step 1, `Enable the restConnector`-->`Enable the Admin REST Connector feature`

the only other question I have is how to provide more detail about accessing the connector- can we provide the JMX endpoints for HTTP calls?

**Comment by dmuelle** [comment]
It would be good to provide more detail about step 3. Particularly about how to access the rest connector with the methods we suggest- 
- what are the JMX endpoints that someone would use for an HTTPS call?
- the IBM docs topic explains how to use jConsole, but Im sure sure how current or relevant the info is. That tool is not OL, but java, so would prefer not to document 3rd party product- might or might not need to bring that info over
- using a client application: what info does someone need if they already have a java client app and want to access JMX from an OL server?

@idlewis can you provide any more info on these points? What makes sense to document?  Thanks

**Comment by dmuelle** [comment]
Hi @idlewis - I've picked up this issue as Manasi is no longer with IBM.

I've revised her draft to remove the security information, which is already documented elsewhere and now linked to from this topic rather than included. I also incorporated more information from [the Liberty IBM Docs topic that was the original source for this draft](https://www.ibm.com/docs/en/was-liberty/nd?topic=jmx-configuring-secure-connection-liberty). When you have a chance, can you review the updated draft and let me know if anything is missing, inaccurate or otherwise unclear?

https://docs-draft-openlibertyio.mybluemix.net/docs/latest/configuring-jmx-connection.html

If any further edits are needed, just let me know. If not, you can sign off by adding the `technical reviewed` label to this issue and the topic will publish with the next docs release. Thanks!

**Comment by idlewis** [comment]
@dmuelle 
I think this is mostly fine. I can't really comment on what is missing, that depends on the purpose of the article, but it seems okay to me.
A couple of minor comments:

"Jakarta Management Extensions (JMX)" - I believe this is still Java Management Extensions, which is confusingly, part of Jakarta Management

I found this sentence a bit confusing:
"If you plan to use the jConsole tool that the trust store contains the certificate of the server in the jConsole connection."
There might be a couple of missing words? Also, I don't think it describes very well which trust store needs to contain the Open Liberty's server certificate.

"Ensure that the CN value of the certificate subjectDN entry is the host name of the machine where the server is running"
Unless this is different to normal security practice (which I don't think it is?) then this sentence seems a bit out of place in this article.

**Comment by dmuelle** [comment]
Thanks for reviewing @idlewis . I made the following changes per your review:

https://docs-draft-openlibertyio.mybluemix.net/docs/21.0.0.12/configuring-jmx-connection.html

- [x] "Jakarta Management Extensions (JMX)" - I believe this is still Java Management Extensions **fixed**
- [x] "If you plan to use the jConsole tool that the trust store contains the certificate of the server in the jConsole connection." There might be a couple of missing words? **fixed**
- [x] "Ensure that the CN value of the certificate subjectDN entry is the host name of the machine where the server is running"
Unless this is different to normal security practice (which I don't think it is?) then this sentence seems a bit out of place in this article.  **removed**

As to the purpose of the article, as far as I understand it, I think it should tell a user how to configure their server to get a secure connection to collect JMX monitoring data- because they are working within an environment that needs to use JMX. The topic doesn't cover how to develop a Java Client or use the JConsole tool, as those tasks aren't specific to Liberty. Though we could develop that information in another topic if I'm wrong about that.

If you see anything else that needs an edit, let me know. If not, you you can sign off by adding the `technical reviewed` label to this issue and the topic will publish with the next docs release. Thanks!

**Comment by idlewis** [comment]
"The topic doesn't cover how to develop a Java Client or use the JConsole tool, as those tasks **aren't specific to Liberty**"
This is broadly true... However, when writing a Java client there are a few Liberty specific parameters. The old docs had an article with some sample code to describe this:
https://www.ibm.com/docs/en/was-liberty/base?topic=liberty-developing-jmx-java-client
Under point 4, there are 2 bullets with a bit of info on Jconsole, and a REST client. Should there be a third bullet linking/describing the Java client as in the above link?

**Comment by dmuelle** [comment]
yes- I think so, thanks for pointing this out. I'm going to add that info to the draft and see how it looks. It may be that it should be in a topic all its own but to start let's try it here and see how it looks. I'll send the updated draft once it's built.

**Comment by dmuelle** [comment]
Hi @idlewis - I've updated the draft to add the JMX client info:

https://draft-openlibertyio.mybluemix.net/docs/21.0.0.12/configuring-jmx-connection.html

I have a few questions to clarify some of the info from the old topic. I'd like to improve the flow of information and, if possible, streamline it. The old topic didn't provide much context and some of the details seem confusing or out of order. If any of the config here represents edge or corner cases that are not likely to be relevant to most users, I think we should remove it and focus on the most concise and simple way to config the client.

> Add the Liberty REST connector client library to the application classpath: `liberty_home/clients/restConnector.jar`

Is that an example of what you would add to the classpath? or an example of the classpath with `restConnector.jar`
 added? Where do you add this code?

> If the server generates the server keystore for a non-production environment, use that keystore as the application client truststore. 

Is this a recommendation for the value of `<truststore location>` in the sample app?

> Set the hostname and port number for the server in the JMXServiceURL constructor by using one of the following options: Use the HTTPS port number. The default number is 9443.
 
How does this first option map to the example? It's confusing because it just tells you to use the HTTPS port number, but not how to specify it. The second option maps directly to the example. Are these two distinct options?

> pass the property as a new environment when you create the JMX connection

Should that be "a new environment _variable_" ?

> The Open Liberty REST connector can use a specific custom SSL socket factory to obtain sockets

This step from the original doc seems out of context. When would a user encounter this exception? How does the example code relate to the previous sample app? Do you create this code as an alternative to or in addition to the previously provided example app?

**Comment by idlewis** [comment]
> Is that an example of what you would add to the classpath? or an example of the classpath with restConnector.jar
> added? Where do you add this code?

How to configure the classpath is very application specific, which is why this statement is vague. Most application developers ought to know how to add the restConnector.jar file to the class path for their application. This is _not_ an example class path. Maybe it would be a bit clearer as something like:
`Add the Liberty REST connector client library (${liberty_home}/clients/restConnector.jar) to the application classpath `

**Comment by idlewis** [comment]
> Is this a recommendation for the value of <truststore location> in the sample app?

Sort of... This one isn't totally liberty specific. The client needs to know the location of a trust store with the correct certificate so that it can trust the server. That location could be anywhere, as the location in your liberty server is configurable.

The example given describes where the default location of the truststore is, if you have allowed the Liberty server to generate its own trust store. So this is a likely example of what a developer would use if they have local, personal Liberty server instance.

**Comment by idlewis** [comment]
> Are these two distinct options?

Err... The original doc is pretty confusing here, and most of it isn't liberty specific. I think `b. Use the contents of the file ${server.output.dir}/logs/state/com.ibm.ws.jmx.rest.address....` is the important part, as that is liberty specific. I'd leave out option a, I think option b is what most people will use to get started, and will then be able to navigate the javadoc for JMXServiceURL if they want to do something different.

**Comment by idlewis** [comment]
> Should that be "a new environment variable" ?

No. The environment is the properties map that is passed to the method that creates the connection. I think the original article is trying to bebrief, and is a bit confusing as a result.

The example passes in an environment to the create the `connector` object and then calls `connector.connect()`. There is an alternative form of the connect method which takes a 2nd environment map: `connector.connect(Map<String,?> env)`. This second form can be used to override properties on a per connection basis.

The original article is trying to describe these two alternatives. In addition, although the property itself is Liberty specific, how you apply the property to JMX connections is generic and not specific to liberty, so describing how to apply the property might be unnecessary.

**Comment by idlewis** [comment]
>This step from the original doc seems out of context.

I think this step is relevant, although not well described. The point here is that a developer can use a custom SSL context for their JMX connections. This allows them to use certificates from a trust store that is specific to their JMX connection. The main example at the top of the article assumes that the developer is using the single trust store that applies to all SSL connections by default. If the default trust store doesn't contain the required certificates for the JMX connection, then a developer will see the exception described. So if they see that exception, they may want to follow the advice here.

However, most of the example code in this section is not Liberty specific. Assuming that a developer has a valid custom SSL context, then the vital (liberty specific part) of this advice is:
`environment.put(ConnectorSettings.CUSTOM_SSLSOCKETFACTORY, sslContext.getSocketFactory());`
So the message is really, 'If you need to use a non-default trust store, then you can use a custom SSL context to enable that, and this property is how you tell the JMX connector to use your custom SSL context'.

**Comment by dmuelle** [comment]
Thanks, @idlewis I've updated the draft per your comments. When you have a chance, let me know if it needs any further edits. If not, you can add the `Developer reviewed` label to this issue and the topic will publish with the 21.0.0.12 docs release. Thanks

https://draft-openlibertyio.mybluemix.net/docs/21.0.0.12/configuring-jmx-connection.html

**Comment by idlewis** [comment]
The edits looks good. I have a couple of minor comments, but don't think they are blockers, so have added the review label.
1. `To run the client application on a different host machine from the server....` This section is generic to any Liberty application, I'm not sure there is any value including it here
2. `Set the following system property in your environment map....` It isn't a 'system property' and as system properties are specific thing in Java, I think this is misleading. Maybe just 'Set the following value...' would do?

**Comment by dmuelle** [comment]
Thanks for reviewing- I've made those changes and the topic will publish with the 21.0.0.12 release

**Comment by ramkumar-k-9286** [comment]
# Peer Review

You can configure a secure JMX connection by enabling the Admin REST connector feature and 
--->
remove linking from `feature` restrict it to `Admin REST connector` 

By default, the Admin REST Connector feature enables the Transport Security feature, which supports TLS connections. 
--->
remove linking from `feature` restrict it to `Transport Security`

Enable the Admin REST connector feature in your server.xml file.
--->
remove linking from `feature` restrict it to `Admin REST connector` 

ensure that the truststore contains the certificate of the server in the jConsole connection.
--->
is truststore a single word?

by using the JConsole tool that is provided by the Java SDK, 
--->
remove linking from `tool` restrict it to `JConsole`

**Comment by dmuelle** [comment]
Thanks Ram- made all changes except `truststore`, which should be one word

https://draft-openlibertyio.mybluemix.net/docs/21.0.0.12/configuring-jmx-connection.html

**Comment by ramkumar-k-9286** [comment]
# Peer Review

Replace the `userid` and `password` values for the `JMXConnector.CREDENTIALS` property. Specify a user ID and password for the administrator or reader role that you configured for step 3 of the previous procedure.
--->
Why is `userid` mentioned in 2 ways in these sentences? Should both `userid` and `password` be mentioned in the same way --> `userid` and `password` in both cases? with monospace as well?

In the following example, the `com.ibm.ws.jmx.connector.client.rest.maxServerWaitTime` property sets the amount of time that the client waits for the server to become available to 6500 milliseconds. The `com.ibm.ws.jmx.connector.client.rest.notificationDeliveryIntervalproperty` sets the maximum amount of time that the server waits for new notifications before it responds to a request for notifications from the client to zero milliseconds.
--->
In one example 6500 milliseconds is used other zero milliseconds. --> should this be 6500 milliseconds and 0 milliseconds?
Also is the property in `com.ibm.ws.jmx.connector.client.rest.notificationDeliveryIntervalproperty` included in monospace by mistake?

If the specified truststore doesn’t contain the required certificates for the JMX connection, the following exception is displayed.
--->
If the specified truststore does not contain the required certificates for the JMX connection, the following exception is displayed.

In this case, you can create your own SSLContext from your own keystores and then use the SocketFactory from that context with the Open Liberty REST connector.
--->
should this be -->In this case, you can create your own `SSLContext` from your own keystores and then use the `SocketFactory` from that context with the Open Liberty REST connector.

The certificates that are installed with Liberty might not contain the host name of where the server is running. If you want to disable hostname verification of SSL certificates, you can set the `com.ibm.ws.jmx.connector.client.disableURLHostnameVerification` system property to `true`, which disables host name verification for all connections.
--->
Hostname mentioned in 2 ways:  host name and  hostname --> i think hostname is the correct usage


Regards,
Ramkumar

**Comment by dmuelle** [comment]
Thanks @ramkumar-k-9286 - addressed your comments in a new update:
https://draft-openlibertyio.mybluemix.net/docs/21.0.0.12/configuring-jmx-connection.html

- [ ] Why is `userid` mentioned in 2 ways in these sentences? Should both `userid` and `password` be mentioned in the same way --> `userid` and `password` in both cases? with monospace as well?  **I don't think so- in the first case, it refers to the actual values that are in the code example, and hence uses monospace. The second case refers to the entities that these values represent, ie a user ID and password for a user who is in the administrator or reader role**

- [x] In one example 6500 milliseconds is used other zero milliseconds. --> should this be 6500 milliseconds and 0 milliseconds? **fixed**

- [x] Also is the property in `com.ibm.ws.jmx.connector.client.rest.notificationDeliveryIntervalproperty` included in monospace by mistake? **fixed**


- [ ] "doesn't" is ok


- [x] should this be -->In this case, you can create your own `SSLContext` from your own keystores and then use the `SocketFactory` from that context with the Open Liberty REST connector. **fixed**

- [x] Hostname mentioned in 2 ways:  host name and  hostname --> i think hostname is the correct usage **fixed**

**Comment by ramkumar-k-9286** [comment]
# Peer Review
LGTM. Proceeding to add the `Peer Reviewed` label.

Regards,
Ramkumar

**Comment by dmuelle** [comment]
Topic is in the vNExt branch and will publish with the 21.0.0.12 docs release
